File: 042. Trapping Rain Water/solution1-n.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    if (height === null || height.length === 0) {
        return 0;
    }
    
    var left = 0;
    var right = height.length - 1;
    var area = 0;
    
    while (left < right) {
        var minHeight = Math.min(height[left], height[right]);
        
        if (minHeight === height[left]) {
            left++;
            while (left < right && height[left] < minHeight) {
                area += minHeight - height[left];
                left++;
            }
        } else {
            right--;
            while (left < right && height[right] < minHeight) {
                area += minHeight - height[right];
                right--;
            }
        }
    }
    
    return area;
};

File: Leet-042. Trapping Rain Water/solution1-n.java
class Solution {
    public int trap(int[] height) {
        if (height == null || height.length <= 2) {
            return 0;
        }
        
        int left = 0, right = height.length - 1;
        int res = 0;
        while (left < right) {
            int h = Math.min(height[left], height[right]);
            if (h == height[left]) {
                while (left < right && height[left] <= h) {
                    res += h - height[left];
                    left++;
                }
            } else {
                while (right > left && height[right] <= h) {
                    res += h - height[right];
                    right--;
                }
            }
        }
        
        return res;
    }
}